File: ts/src/engine/manifest.ts
/**
 * Manifest model + parsing, ported from manifest.py, plus the auto-manifest
 * used by zero-config `quorable review` (Blocker 2 in the plan): the target
 * file becomes the tier-1 primary and `--context` directories glob into
 * tier-2 entries routed to stage1. A hand-written manifest still wins when
 * present.
 */

import fs from "node:fs";
import path from "node:path";

import { parse as parseYaml } from "yaml";

export interface ManifestEntry {
  name: string;
  path: string;
  format: string;
  role: string;
  tier: 1 | 2 | 3;
  sendTo: string[];
  critical: boolean;
  notes: string;
}

/** A parsed input document ready for inclusion in LLM prompts. */
export interface DocumentModel {
  name: string;
  role: string;
  tier: number;
  content: string;
  pageCount: number;
  charCount: number;
  sha256: string;
  truncated: boolean;
}

const FORMAT_BY_EXT: Record<string, string> = {
  ".pdf": "pdf",
  ".md": "markdown",
  ".markdown": "markdown",
  ".txt": "markdown",
  ".yaml": "yaml",
  ".yml": "yaml",
  ".docx": "docx",
};

export function guessFormat(p: string): string {
  return FORMAT_BY_EXT[path.extname(p).toLowerCase()] ?? "markdown";
}

interface RawEntry {
  path: string;
  format?: string;
  role?: string;
  tier?: number;
  send_to?: string[];
  critical?: boolean;
  notes?: string;
}

function isEntryDict(val: unknown): val is RawEntry {
  return typeof val === "object" && val !== null && "path" in val;
}

function parseSingle(raw: RawEntry, inputsDir: string, fallbackName: string): ManifestEntry {
  const absPath = path.resolve(inputsDir, raw.path);
  return {
    name: fallbackName,
    path: absPath,
    format: raw.format ?? guessFormat(raw.path),
    role: raw.role ?? "",
    tier: (raw.tier ?? 1) as 1 | 2 | 3,
    sendTo: raw.send_to ?? [],
    critical: raw.critical ?? false,
    notes: raw.notes ?? "",
  };
}

function parseList(items: unknown[], inputsDir: string): ManifestEntry[] {
  const entries: ManifestEntry[] = [];
  for (const item of items) {
    if (!isEntryDict(item)) continue;
    const name = path.parse(item.path).name;
    entries.push(parseSingle(item, inputsDir, name));
  }
  return entries;
}

/**
 * Parse a manifest YAML into a flat entry list. Sections are free-form: any
 * top-level key may hold a single entry, a dict of named entries, a list of
 * entries, or a dict of named lists.
 */
export function loadManifest(manifestPath: string, inputsDir: string): ManifestEntry[] {
  const raw = parseYaml(fs.readFileSync(manifestPath, "utf-8")) as Record<
    string,
    unknown
  > | null;
  const entries: ManifestEntry[] = [];

  for (const [key, val] of Object.entries(raw ?? {})) {
    if (val === null || val === undefined) continue;
    if (isEntryDict(val)) {
      entries.push(parseSingle(val, inputsDir, key));
    } else if (typeof val === "object" && !Array.isArray(val)) {
      for (const [subKey, subVal] of Object.entries(val as Record<string, unknown>)) {
        if (subVal === null || subVal === undefined) continue;
        if (isEntryDict(subVal)) {
          entries.push(parseSingle(subVal, inputsDir, subKey));
        } else if (Array.isArray(subVal)) {
          entries.push(...parseList(subVal, inputsDir));
        }
      }
    } else if (Array.isArray(val)) {
      entries.push(...parseList(val, inputsDir));
    }
  }

  // Enforce that all critical files exist (fail loudly — the grant fork's
  // silent-skip defect is exactly what this guards against).
  for (const entry of entries) {
    if (entry.critical && !fs.existsSync(entry.path)) {
      throw new Error(
        `Critical document missing: ${entry.name} (${entry.path}). ` +
          `The pipeline cannot run without this file.`,
      );
    }
  }

  return entries;
}

/**
 * Declared-vs-loaded assertion (plan M9): every declared, existing manifest
 * entry must appear in the loaded document set — silently skipped context
 * poisons every downstream review.
 */
export function assertManifestLoaded(
  entries: ManifestEntry[],
  documents: Record<string, DocumentModel>,
): void {
  const missing = entries.filter((e) => fs.existsSync(e.path) && !(e.name in documents));
  if (missing.length > 0) {
    throw new Error(
      `Manifest declared ${missing.length} document(s) that failed to load: ` +
        missing.map((e) => e.name).join(", ") +
        ". Refusing to run with silently missing context.",
    );
  }
}

const CONTEXT_EXTENSIONS = new Set([
  ".md",
  ".markdown",
  ".txt",
  ".pdf",
  ".yaml",
  ".yml",
  ".docx",
]);

/**
 * Build an auto-manifest for zero-config review: the target file is the
 * tier-1 primary (critical, never truncated); each --context directory is
 * globbed (recursively) into tier-2 entries routed to stage1 + stage2.
 */
export function autoManifest(
  targetPath: string,
  contextDirs: string[] = [],
  opts: { primaryName?: string } = {},
): ManifestEntry[] {
  const primaryName = opts.primaryName ?? "primary_document";
  const entries: ManifestEntry[] = [
    {
      name: primaryName,
      path: path.resolve(targetPath),
      format: guessFormat(targetPath),
      role: "Primary document under review",
      tier: 1,
      sendTo: ["stage1", "stage2", "stage3"],
      critical: true,
      notes: "",
    },
  ];

  const seen = new Set<string>([path.resolve(targetPath)]);
  for (const dir of contextDirs) {
    for (const file of walkFiles(path.resolve(dir))) {
      if (!CONTEXT_EXTENSIONS.has(path.extname(file).toLowerCase())) continue;
      if (seen.has(file)) continue;
      seen.add(file);
      const base = path.parse(file).name;
      let name = base;
      let suffix = 2;
      while (entries.some((e) => e.name === name)) {
        name = `${base}_${suffix++}`;
      }
      entries.push({
        name,
        path: file,
        format: guessFormat(file),
        role: `Context document (${path.relative(dir, file) || path.basename(file)})`,
        tier: 2,
        sendTo: ["stage1", "stage2"],
        critical: false,
        notes: "",
      });
    }
  }
  return entries;
}

function* walkFiles(dir: string): Generator<string> {
  let items: fs.Dirent[];
  try {
    items = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return;
  }
  const sorted = items.sort((a, b) => a.name.localeCompare(b.name));
  for (const item of sorted) {
    if (item.name.startsWith(".")) continue;
    const full = path.join(dir, item.name);
    if (item.isDirectory()) {
      yield* walkFiles(full);
    } else if (item.isFile()) {
      yield full;
    }
  }
}

File: src/quorable/engine/manifest.py
"""Parse inputs/manifest.yaml into structured ManifestEntry objects.

Forked from the reference implementation's manifest module and genericized: the parent hardcoded its
legal section keys (core / rulings / defense / plaintiff / evidence / judge).
Here sections are free-form — any top-level key works, so each project can
organize its manifest around its own domain. The ManifestEntry shape, tiers,
and `send_to` routing are unchanged.

Supported shapes under any top-level key:

- a single entry dict with a `path` key (entry named after the top-level key),
- a dict of named entry dicts (entries named after the sub-keys),
- a list of entry dicts (entries named after each path's stem),
- a dict of named lists (one level of nesting, e.g. the parent's
  defense.primary/replies/historical — entries named after path stems).

`system_prompt` is just another single-entry key by this scheme, preserving
the parent's manifest format.
"""
from __future__ import annotations

import logging
from pathlib import Path

import yaml
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)


class ManifestEntry(BaseModel):
    """One document described in the manifest."""

    name: str
    path: Path
    format: str = "markdown"
    role: str = ""
    tier: int = Field(ge=1, le=3)
    send_to: list[str] = Field(default_factory=list)
    critical: bool = False
    notes: str = ""


def _parse_single(raw: dict, inputs_dir: Path, fallback_name: str) -> ManifestEntry:
    """Turn one manifest entry dict into a ManifestEntry."""
    rel_path = raw["path"]
    abs_path = inputs_dir / rel_path
    if not abs_path.exists():
        logger.warning("Manifest references missing file: %s", abs_path)

    return ManifestEntry(
        name=fallback_name,
        path=abs_path,
        format=raw.get("format", _guess_format(rel_path)),
        role=raw.get("role", ""),
        tier=raw.get("tier", 1),
        send_to=raw.get("send_to", []),
        critical=raw.get("critical", False),
        notes=raw.get("notes", ""),
    )


def _guess_format(path_str: str) -> str:
    suffix = Path(path_str).suffix.lower()
    return {".pdf": "pdf", ".md": "markdown", ".yaml": "yaml", ".yml": "yaml"}.get(
        suffix, "markdown"
    )


def _is_entry_dict(val: object) -> bool:
    return isinstance(val, dict) and "path" in val


def _parse_list(items: list, inputs_dir: Path) -> list[ManifestEntry]:
    entries: list[ManifestEntry] = []
    for item in items:
        if not _is_entry_dict(item):
            logger.warning("Skipping malformed manifest list item: %r", item)
            continue
        name = Path(item["path"]).stem
        entries.append(_parse_single(item, inputs_dir, name))
    return entries


def load_manifest(manifest_path: Path, inputs_dir: Path) -> list[ManifestEntry]:
    """Parse the manifest YAML and return a flat list of ManifestEntry."""
    with open(manifest_path) as f:
        raw = yaml.safe_load(f) or {}

    entries: list[ManifestEntry] = []

    for key, val in raw.items():
        if val is None:
            continue
        if _is_entry_dict(val):
            # Single entry named after the top-level key (e.g. system_prompt)
            entries.append(_parse_single(val, inputs_dir, key))
        elif isinstance(val, dict):
            # Dict of named entries, or dict of named lists
            for sub_key, sub_val in val.items():
                if sub_val is None:
                    continue
                if _is_entry_dict(sub_val):
                    entries.append(_parse_single(sub_val, inputs_dir, sub_key))
                elif isinstance(sub_val, list):
                    entries.extend(_parse_list(sub_val, inputs_dir))
                else:
                    logger.warning(
                        "Skipping malformed manifest section %s.%s", key, sub_key,
                    )
        elif isinstance(val, list):
            entries.extend(_parse_list(val, inputs_dir))
        else:
            logger.warning("Skipping malformed manifest section %s", key)

    logger.info(
        "Loaded manifest with %d entries (%d tier-1, %d tier-2, %d tier-3)",
        len(entries),
        sum(1 for e in entries if e.tier == 1),
        sum(1 for e in entries if e.tier == 2),
        sum(1 for e in entries if e.tier == 3),
    )

    # Enforce that all critical files exist
    for entry in entries:
        if entry.critical and not entry.path.exists():
            raise FileNotFoundError(
                f"Critical document missing: {entry.name} ({entry.path}). "
                f"The pipeline cannot run without this file."
            )

    return entries
